src/V4Router.sol

```diff
@@ -0,0 +1,179 @@
+                revert UnsupportedAction(action);
+            }
+        } else {
+            if (action == Actions.SETTLE) {
+                // equivalent: abi.decode(params, (Currency))
+                Currency currency;
+                assembly ("memory-safe") {
+                    currency := calldataload(params.offset)
+                }
+
+                // TODO support address(this) paying too
+                _payAndSettle(currency, _msgSender());
+            } else if (action == Actions.TAKE) {
+                // equivalent: abi.decode(params, (Currency, address))
+                Currency currency;
+                address recipient;
+                assembly ("memory-safe") {
+                    currency := calldataload(params.offset)
+                    recipient := calldataload(add(params.offset, 0x20))
+                }
+
+                // TODO add min amount??
+                _take(currency, recipient);
+            } else {
+                revert UnsupportedAction(action);
+            }
+        }
+    }
+
+    function _swapExactInputSingle(IV4Router.ExactInputSingleParams memory params) private {
+        _swap(
+            params.poolKey,
+            params.zeroForOne,
+            int256(-int128(params.amountIn)),
+            params.sqrtPriceLimitX96,
+            params.hookData
+        );
+    }
+
+    function _swapExactInput(IV4Router.ExactInputParams memory params) private {
+        unchecked {
+            uint256 pathLength = params.path.length;
+            uint128 amountOut;
+
+            for (uint256 i = 0; i < pathLength; i++) {
+                (PoolKey memory poolKey, bool zeroForOne) = _getPoolAndSwapDirection(params.path[i], params.currencyIn);
+                amountOut =
+                    uint128(_swap(poolKey, zeroForOne, int256(-int128(params.amountIn)), 0, params.path[i].hookData));
+
+                params.amountIn = amountOut;
+                params.currencyIn = params.path[i].intermediateCurrency;
+            }
+
+            if (amountOut < params.amountOutMinimum) revert TooLittleReceived();
+        }
+    }
+
+    function _swapExactOutputSingle(IV4Router.ExactOutputSingleParams memory params) private {
+        _swap(
+            params.poolKey,
+            params.zeroForOne,
+            int256(int128(params.amountOut)),
+            params.sqrtPriceLimitX96,
+            params.hookData
+        );
+    }
+
+    function _swapExactOutput(IV4Router.ExactOutputParams memory params) private {
+        unchecked {
+            uint256 pathLength = params.path.length;
+            uint128 amountIn;
+
+            for (uint256 i = pathLength; i > 0; i--) {
+                (PoolKey memory poolKey, bool oneForZero) =
+                    _getPoolAndSwapDirection(params.path[i - 1], params.currencyOut);
+                amountIn = uint128(
+                    -_swap(poolKey, !oneForZero, int256(int128(params.amountOut)), 0, params.path[i - 1].hookData)
```

Comment:
why is this safe to flip the sign? is it guaranteed to be negative? its just not super obvious

Reply:
so amountSpecified is positive (because params.amountOut is positive), which makes it exactOut. Which means itll be swapped and generate an input delta, which will be negative. WIll leave a comment

Reply:
great! yeah comment would be helpful its hard to remember all the signs for specific swap types